### PR TITLE
API:  Make .shift always copy (Fixes #22397)

### LINKED ISSUE #22397: Make Series.shift always a copy?
Right now, `Series.shift(0)` will just return the series. Shifting for all other periods induces a copy:

```python
In [1]: import pandas as pd

In [2]: a = pd.Series([1, 2])

In [3]: a.shift(1) is a
Out[3]: False

In [4]: a.shift(0) is a
Out[4]: True
```

Should we defensively copy on `0` as well, for a consistent user experience?

### PR DESCRIPTION
**Pull request by AaronCritchley**
- [x] closes #22397
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/generic/test_series.py [hunk 1]
```diff
@@ -227,3 +227,9 @@ def test_valid_deprecated(self):
         # GH18800
         with tm.assert_produces_warning(FutureWarning):
             pd.Series([]).valid()
+
+    def test_shift_always_copy(self):
+        # GH22397
+        s = Series([1, 2, 3])
+        assert s.shift(0) is not s
+        assert s.shift(1) is not s
```
Comment (gfyoung) [inline pandas/tests/generic/test_series.py]: Can we parametrize this?  (I know it's only two examples, but why not three for example?)

Comment (TomAugspurger) [reply pandas/tests/generic/test_series.py]: This seems fine to be as is.

**Review by TomAugspurger — COMMENTED** [review]
LGTM.

We should confirm that we're OK with outright changing the behavior, rather than deprecating it first. Deprecation would require adding a keyword that would then become obsolete, which I would rather avoid.

File: pandas/tests/generic/test_series.py [hunk 2]
```diff
@@ -227,3 +227,9 @@ def test_valid_deprecated(self):
         # GH18800
         with tm.assert_produces_warning(FutureWarning):
             pd.Series([]).valid()
+
+    def test_shift_always_copy(self):
+        # GH22397
+        s = Series([1, 2, 3])
```
Comment (jreback) [inline pandas/tests/generic/test_series.py]: can u also test for datetime types

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/generic/test_series.py (hunk 1 above)
Comment (AaronCritchley) [reply pandas/tests/generic/test_series.py]: Done

File: pandas/tests/generic/test_series.py (hunk 2 above)
Comment (AaronCritchley) [reply pandas/tests/generic/test_series.py]: Done (assuming DatetimeIndex was meant, which required another fix) if you just meant a series of datetimes, will change - let me know

File: pandas/tests/generic/test_series.py [hunk 3]
```diff
@@ -227,3 +227,12 @@ def test_valid_deprecated(self):
         # GH18800
         with tm.assert_produces_warning(FutureWarning):
             pd.Series([]).valid()
+
+    @pytest.mark.parametrize("s", [
+        Series([np.arange(5)]),
```
Comment (jreback) [inline pandas/tests/generic/test_series.py]: can you add period & timedelta here as well :<

Comment (jreback) [reply pandas/tests/generic/test_series.py]: we dont really have a great place for these series / index tests functionailty, except for test_base which is already too big.

cc @jbrockmendel @mroeschke

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/generic/test_series.py (hunk 3 above)
Comment (AaronCritchley) [reply pandas/tests/generic/test_series.py]: ofc, will do asap, thanks for the input

Comment (AaronCritchley) [reply pandas/tests/generic/test_series.py]: Sorry for the delay on this, but can you please clarify? I wasn't sure if it was expected that we can shift Period / Timedelta's or if we're supposed to shift by them, things I've tested below:

Shifting on them:
```python
>>> pd.Timedelta('1 days 2 hours').shift
Traceback (most recent call last):
  File "/anaconda3/envs/pandas-dev/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2963, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-17-fc984c065985>", line 1, in <module>
    pd.Timedelta('1 days 2 hours').shift
AttributeError: 'Timedelta' object has no attribute 'shift'
>>> pd.Period('2011-01', 'M').shift
Traceback (most recent call last):
  File "/anaconda3/envs/pandas-dev/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2963, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-18-d3c2e47fd2e3>", line 1, in <module>
    pd.Period('2011-01', 'M').shift
AttributeError: 'Period' object has no attribute 'shift'
```

Shifting with them:
```python
>>> pd.date_range('1/1/2011', periods=24, freq='H').shift(pd.Timedelta('1 days 2 hours'))
Traceback (most recent call last):
  File "/anaconda3/envs/pandas-dev/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2963, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-19-f8499141727f>", line 1, in <module>
    pd.date_range('1/1/2011', periods=24, freq='H').shift(pd.Timedelta('1 days 2 hours'))
  File "/anaconda3/envs/pandas-dev/lib/python3.6/site-packages/pandas/core/indexes/datetimelike.py", line 784, in shift
    start = self[0] + n * self.freq
  File "/anaconda3/envs/pandas-dev/lib/python3.6/site-packages/pandas/tseries/offsets.py", line 425, in __rmul__
    return self.__mul__(someInt)
  File "/anaconda3/envs/pandas-dev/lib/python3.6/site-packages/pandas/tseries/offsets.py", line 422, in __mul__
    **self.kwds)
  File "/anaconda3/envs/pandas-dev/lib/python3.6/site-packages/pandas/tseries/offsets.py", line 195, in __init__
    self.n = int(n)
TypeError: int() argument must be a string, a bytes-like object or a number, not 'Timedelta'
>>> pd.date_range('1/1/2011', periods=24, freq='H').shift(pd.Period('2011-02', 'M'))
Traceback (most recent call last):
  File "/anaconda3/envs/pandas-dev/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2963, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-24-b87cfa9cba95>", line 1, in <module>
    pd.date_range('1/1/2011', periods=24, freq='H').shift(pd.Period('2011-02', 'M'))
  File "/anaconda3/envs/pandas-dev/lib/python3.6/site-packages/pandas/core/indexes/datetimelike.py", line 784, in shift
    start = self[0] + n * self.freq
  File "/anaconda3/envs/pandas-dev/lib/python3.6/site-packages/pandas/tseries/offsets.py", line 425, in __rmul__
    return self.__mul__(someInt)
  File "/anaconda3/envs/pandas-dev/lib/python3.6/site-packages/pandas/tseries/offsets.py", line 421, in __mul__
    return self.__class__(n=someInt * self.n, normalize=self.normalize,
TypeError: unsupported operand type(s) for *: 'Period' and 'int'
```

Genuine apologies if I'm being stupid here!

Comment (TomAugspurger) [reply pandas/tests/generic/test_series.py]: I think the idea is to make a series with a datetimeindex and shift using the `freq` argument.


```python
In [16]: s = pd.Series(range(5), index=pd.date_range("2017", periods=5))

In [17]: s.shift(freq=pd.Timedelta('1D'))
Out[17]:
2017-01-02    0
2017-01-03    1
2017-01-04    2
2017-01-05    3
2017-01-06    4
Freq: D, dtype: int64
```

it'll complicate the test a bit. let us know if you need assistance.

Comment (AaronCritchley) [reply pandas/tests/generic/test_series.py]: Added the test for timedelta, I couldn't find any docs on the intended usage of shifting by pd.Period? I tried a few approaches but couldn't get any to work. Happy to add the test if you could provide a small example!

Comment (TomAugspurger) [reply pandas/tests/generic/test_series.py]: I'm not sure if period is valid there. I only tried timedelta.

**Comment by TomAugspurger** [comment]
You'll need to fetch and merge master for the CI failures to be fixed.

**Review by jreback — APPROVED** [review]

**Pull request merged**